gax-httpjson/src/main/java/com/google/api/gax/httpjson/ApiMessageOperationTransformers.java

```diff
@@ -0,0 +1,122 @@
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.api.core.ApiFunction;
+import com.google.api.core.BetaApi;
+import com.google.api.gax.longrunning.OperationSnapshot;
+import com.google.api.gax.rpc.ApiExceptionFactory;
+import com.google.api.gax.rpc.StatusCode.Code;
+
+/**
+ * Transformers from OperationSnapshot wrappers to the underlying native ApiMessage objects. Public
+ * for technical reasons; intended for use by generated code.
+ */
+@BetaApi("The surface for use by generated code is not stable yet and may change in the future.")
+public class ApiMessageOperationTransformers {
+  private ApiMessageOperationTransformers() {}
+
+  public static class ResponseTransformer<ResponseT extends ApiMessage>
+      implements ApiFunction<OperationSnapshot, ResponseT> {
+    private final Class<ResponseT> responseTClass;
+
+    private ResponseTransformer(Class<ResponseT> responseTClass) {
+      this.responseTClass = responseTClass;
+    }
+
+    /** Unwraps an OperationSnapshot and returns the contained method response message. */
+    @SuppressWarnings("unchecked")
+    public ResponseT apply(OperationSnapshot operationSnapshot) {
+      if (!operationSnapshot.getErrorCode().getCode().equals(Code.OK)) {
+        throw ApiExceptionFactory.createException(
+            "Operation with name \""
+                + operationSnapshot.getName()
+                + "\" failed with status = "
+                + operationSnapshot.getErrorCode()
+                + " and message = "
+                + operationSnapshot.getErrorMessage(),
+            null,
+            operationSnapshot.getErrorCode(),
+            false);
+      }
+      if (!operationSnapshot.getResponse().getClass().isAssignableFrom(responseTClass)) {
```

Comment:
Is `operationSnapshot.getResponse()` returning a Java proxy class? Or in other words, why `isAssignableFrom()` instead of `isInstance()`. `isAssignableFrom()` is heavier.

Reply:
I don't know why `isInstance()` doesn't work; the error says
```
... its response type com.google.api.gax.httpjson.EmptyMessage cannot be cast to com.google.api.gax.httpjson.EmptyMessage.
```

But `isAssignableFrom()` does work for this use case.

Reply:
This is very suspicious, the error message says it cannot cast class to itself. Something is wrong here.

Reply:
i know! i looked on stackoverflow and some answers say that this might happen if you have different classloaders? idk